### PR TITLE
Support for readOnly realms

### LINKED ISSUE #392: Support read only realms
Needed to support bundled realm files http://stackoverflow.com/questions/36682944/is-there-a-way-to-pre-package-realm-database-default-realm-through-react-nativ

### PR DESCRIPTION
**Pull request by alazier**
Closes #392

### CODE REVIEW

File: src/js_realm.hpp [hunk 1]
```diff
@@ -267,28 +263,36 @@ void Realm<T>::constructor(ContextType ctx, ObjectType this_object, size_t argc,
...
             ValueType version_value = Object::get_property(ctx, object, schema_version_string);
             if (!Value::is_undefined(ctx, version_value)) {
                 config.schema_version = Value::validated_to_number(ctx, version_value, "schemaVersion");
             }
             else {
                 config.schema_version = 0;
             }
-            
+
+	    	static const String migration_string = "migration";
```
Comment (appden) [inline src/js_realm.hpp]: Over-indented

File: tests/js/realm-tests.js [hunk 2]
```diff
@@ -119,6 +119,30 @@ module.exports = BaseTest.extend({
...
+        TestCase.assertEqual(realm.readOnly, true);
+
+        TestCase.assertThrows(function() {
+            realm.write(function() {});
+        });
+        realm.close();
+
+        realm = new Realm({readOnly: true});
+        TestCase.assertEqual(realm.schema.length, 1);
+        TestCase.assertEqual(realm.readOnly, true);
```
Comment (appden) [inline tests/js/realm-tests.js]: Does can you re-open a read-only realm without it being read-only, and vice versa?

Comment (alazier) [inline tests/js/realm-tests.js]: you should be able to. readonly realms are not cached

Comment (appden) [inline tests/js/realm-tests.js]: Should we test this?

File: lib/browser/index.js [hunk 3]
```diff
@@ -71,6 +71,8 @@ export default class Realm {
 
         [
             'path',
+            'readOnly',
```
Comment (appden) [inline lib/browser/index.js]: I think `readonly` would be nicer IMHO

Comment (alazier) [inline lib/browser/index.js]: I'm on the fence but all the other bindings use the current capitalization.

File: tests/js/realm-tests.js (hunk 2 above)
Comment (alazier) [inline tests/js/realm-tests.js]: yeah

File: lib/browser/index.js (hunk 3 above)
Comment (appden) [inline lib/browser/index.js]: I understand, but I think we should do what feels right for our binding. If you feel strongly we should keep it this way, then that's fine.

**Comment by appden** [comment]
Looks good :+1:

**Comment by chaodoze** [comment]
Can you provide sample code and steps on how to create a pre-packaged read only realm, that can be bundled in react-native? Thanks!

**Comment by Bogidon** [comment]
Does readOnly add performance/size optimizations (like removing locks)? Couldn't find a matching PR in [realm/realm-core](https://github.com/realm/realm-core).

**Comment by Bogidon** [comment]
I think my previous comment was unclear. What I mean to ask is this: is there a way to turn a read-write realm it into a read-only realm and in the process remove any indexing/features required for write support? I think a lot of users of `realm-js` are looking to bundle their apps with pre-built databases, which (on iOS at least) must be copied to read-write directories due to OS restrictions. 

It would certainly be beneficial to trim the DB size as much as possible. I'm currently shipping a gzipped `.realm` to App Stores (with surprisingly good results), but I'm sure that's all that can be done.

**Pull request merged**